ambry-replication/src/main/java/com/github/ambry/replication/CloudToStoreReplicationManager.java

```diff
@@ -240,9 +240,14 @@ private void removeCloudReplica(String partitionName) {
   /**
    * Randomly select a {@link DataNodeId} from list of {@code vcrNodes}.
    * @return randomly selected {@link DataNodeId} object.
+   * @throws ReplicationException if there are no vcr nodes.
    */
-  private DataNodeId getCloudDataNode() {
-    return vcrNodes.get().toArray(new CloudDataNode[0])[Utils.getRandomShort(new Random()) % vcrNodes.get().size()];
+  private DataNodeId getCloudDataNode() throws ReplicationException {
+    List<CloudDataNode> nodes = vcrNodes.get();
+    if (nodes.isEmpty()) {
+      throw new ReplicationException("No VCR node found to replicate partition from cloud.");
+    }
+    return nodes.get(Utils.getRandomShort(new Random()) % nodes.size());
```

Comment:
I didn't notice it before, but I think initializing a new Random() each time will result in the same value each time.  At least that's how it used to behave, not sure if it's still the case.  Either way, I would initialize the Random statically and reuse.  I recommend adding a test case that calls this method consecutive times and verifies different results are returned.  And one to test the no-nodes exception case.

Reply:
Sure, I will change the `Random` to static, and add the tests.
Note that ideally, in a perfect random setting, I assume it could be possible for a few calls to `getNextInt()` to return same value (Maybe this will happen rarely, but this is a possibility). So in my test I am just checking that multiple calls return more than one value.

As for the behavior of `Random`, from the constructor of `Random` it seems like every instantiation gets a unique seed. Below code snippet is from `java.util.Random` class.

```
    /**
     * Creates a new random number generator. This constructor sets
     * the seed of the random number generator to a value very likely
     * to be distinct from any other invocation of this constructor.
     */
    public Random() {
        this(seedUniquifier() ^ System.nanoTime());
    }
```


Also I wrote an independent test that sort of proves this.

```
public class RandomTest {
	public static void main(String[] args) {
		for(int i = 0; i < 10; i++) {
			System.out.println(new Random().nextInt()%17);
		}
	}
}
```

Reply:
My information was from at least 10 years ago, so they likely changed it since then.